IPython/extensions/deduperreload/deduperreload.py

```diff
@@ -0,0 +1,477 @@
+from __future__ import annotations
+import ast
+import contextlib
+import os
+import sys
+import textwrap
+from types import ModuleType
+from typing import TYPE_CHECKING, Any, NamedTuple
+
+from IPython.extensions.deduperreload.deduperreload_patching import (
+    DeduperReloaderPatchingMixin,
+)
+
+if TYPE_CHECKING:
+    from IPython.core.interactiveshell import ExecutionResult
+
+    TDefinitionAst = ast.FunctionDef | ast.AsyncFunctionDef
+
+DefinitionAst = (ast.FunctionDef, ast.AsyncFunctionDef)
+
+
+def get_module_file_name(module: ModuleType | str) -> str:
```

Comment:
As a note (not a request to change). Those `|` fails on older version of Python when using runtime types extraction of type (https://docs.python.org/3/library/typing.html#typing.get_type_hints). Som in some case  `from __future__ import annotations` is not sufficient to use the new annotation syntax.